contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -1639,3 +1646,244 @@ pltsql_replace_non_determinstic(text *src_text, text *from_text, text *to_text,
+										&status);
+		if (U_FAILURE(status))
+			ereport(ERROR,
+					(errcode(ERRCODE_INTERNAL_ERROR),
+						errmsg("failed to perform ICU search: %s",
+							   u_errorName(status))));
+
+		usearch_setAttribute(usearch, USEARCH_OVERLAP, USEARCH_ON, &status);
+
+		/* substr should start matching from the first position in src string */
+		u16_pos = usearch_preceding(usearch, 1, &status);
+		if (!U_FAILURE(status) && u16_pos != USEARCH_DONE)
+		{
+			u8_pos = translate_char_pos(VARDATA_ANY(src_text), src_len_utf8,
+										src_uchar, src_ulen, u16_pos,
+										NULL);
+			matched_length_u16 = usearch_getMatchedLength(usearch);
+			u8_endpos = translate_char_pos(VARDATA_ANY(src_text), src_len_utf8,
+										src_uchar, src_ulen, u16_pos + matched_length_u16,
+										NULL);
+			*matched_len = u8_endpos - u8_pos;
+		}
+		else
+			u8_pos = -1;
+
+		pfree(src_uchar);
+		pfree(substr_uchar);
+		usearch_close(usearch);
+
+		return u8_pos < 0 ? false : true;
+#else
+	ereport(ERROR,
+			(errcode(ERRCODE_EXTERNAL_ROUTINE_EXCEPTION),
+				errmsg("This function requires ICU library, which is not available")));
+#endif
+    }
+
+	return false;
+}
+
+static int
+patindex_ai_match_text(char *input_str, char *pattern, Oid cid)
+{
+	bool start_offset = false;
+	int  itr = 0;
+
+	while (*pattern == '%')
+	{
+		pattern++;
+		start_offset = true;
+	}
+	if (strlen(pattern) == 0)
+		return 1;
+
+	while (*input_str != '\0')
+	{
+		char  *t = input_str;
+		char  *p = pattern;
+		int   tlen = strlen(t),
+		      plen = strlen(pattern);
+
+		itr++;
+
+		while (tlen > 0 && plen > 0)
+		{
+			if (*p == '%')
+			{
+				if (patindex_ai_match_text(t, p, cid))
+					return itr;
+				else
+					return 0;
+			}
+			else if (*p == '_')
+			{
+				/* _ matches any single character, and we know there is one */
+				NextChar(t, tlen);
+				NextByte(p, plen);
+				continue;
+			}
+			else if (*p == '[')
+			{
+				/* Tsql deal with [ and ] wild character */
+				bool find_match = false, reverse_mode = false, close_bracket = false;
+				const char * prev = NULL;
+
+				NextByte(p, plen);
+				if (plen > 0 && *p == '^')
+				{
+					reverse_mode = true;
+					NextByte(p, plen);
+				}
+				while (plen > 0)
+				{
+					if (*p == ']')
+					{
+						close_bracket = true;
+						NextByte(p, plen);
+						break;
+					}
+					if (find_match)
+					{
+						NextByte(p, plen);
+						continue;
+					}
+					if (*p == '-' && prev)
+					{
+						NextByte(p, plen);
+						Assert(cid != InvalidOid);
+						if (varstr_cmp(t, pg_mblen(t), prev, pg_mblen(prev), cid) >= 0 && varstr_cmp(t, pg_mblen(t), p, pg_mblen(p), cid) <= 0)
+						{
+							find_match = true;
+						}
+						prev = NULL;
+						NextByte(p, plen);
+					}
+					else
+					{
+						int p_start_len = plen, matched_idx = 0;
+						char *p_start = p;
+						text *src_text, *substr_text;
+						prev = p;
+
+						/* find the string till the next special character */
+						while (plen > 0 && *p != ']' && *p != '-')
```

Comment:
If we have pattern like `[abcm-o]` then character `m` should not be considered along with `abc`, right? because it is part of range

Reply:
The first character of range is actually considered but not the 2nd one.
```
1> SELECT 1 WHERE 'Æ' like '[A-AE]' COLLATE Latin1_General_CI_AI
2> go           
-----------


1> SELECT 1 WHERE 'Æ' like '[AE-E]' COLLATE Latin1_General_CI_AI
2> go
-----------
          1
```